{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 1  \n",
    "---  \n",
    "## Demultiplexing:   \n",
    "---  \n",
    "### 1) File Identification  \n",
    "\n",
    "| File Name                    | ID      |  \n",
    "|:----------------------------:|:-------:|  \n",
    "| 1294_S1_L008_R1_001.fastq.gz | read1  |  \n",
    "| 1294_S1_L008_R2_001.fastq.gz | index1 |  \n",
    "| 1294_S1_L008_R3_001.fastq.gz | index2 |  \n",
    "| 1294_S1_L008_R4_001.fastq.gz | read2  |  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2a) Python Code for Mean Q-score Analysis  \n",
    "---  \n",
    "Written by: Max Hills  \n",
    "**See Mean Quality-Score Graphs and Tables Below**   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "% matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "import numpy as np\n",
    "def convert_phred(letter):\n",
    "    \"\"\"The function takes a single string character and converts it into the Illumina quality score.\"\"\"\n",
    "    phred_score = ord(letter)-33\n",
    "    return phred_score\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "def avg_qs(fastq):\n",
    "    \"\"\"INPUT: fastq - A standard format fastq file of type str, readlen - The int length of the reads (in nucleotides)\n",
    "    OUTPUT: A table of average q-scores at each position, A line-graph of average q-scores at each position\"\"\"\n",
    "    with open(fastq,'r') as f:\n",
    "        # Initialize a line-count.\n",
    "        LC = 0\n",
    "        # Initialize a read-count.\n",
    "        RC = 0\n",
    "        # Iterate through each line in the file.\n",
    "        for line in f:        \n",
    "            # Increment the line count.\n",
    "            LC += 1        \n",
    "            # If the line number is a multiple of 4, do the following...\n",
    "            if LC%4 == 0:\n",
    "                line = line.strip()\n",
    "                readlen = len(line)\n",
    "                if RC == 0:\n",
    "                    all_qscores = np.zeros(readlen,dtype=int)\n",
    "                # Increment the read-count.\n",
    "                RC += 1\n",
    "                # Initialize charIndex\n",
    "                charIndex = 0\n",
    "                for char in line:\n",
    "                    # Get the decimal phred score at this position.\n",
    "                    score = convert_phred(char)\n",
    "                    # Store the score for the current position at its index within all_qscores. \n",
    "                    all_qscores[charIndex] += score\n",
    "                    # Increment charIndex.\n",
    "                    charIndex += 1\n",
    "    mean_scores = all_qscores/RC\n",
    "    table = fastq.replace('fastq','txt').replace('.gz','')\n",
    "    graph = fastq.replace('fastq','png').replace('.gz','')\n",
    "    with open(table, 'w') as t:\n",
    "        # Print a header for the columns.\n",
    "        print(\"# Base Pos\\tMean Quality Score\",file=t)\n",
    "        print(\"===========\\t==================\",file=t)\n",
    "        for i in range(readlen):\n",
    "            template = \"{}\\t\\t{:0.1f}\"\n",
    "            out = str(template.format(i,mean_scores[i]))\n",
    "            print(out,file=t)\n",
    "    # Initialize empty lists to hold x,y axis labels.\n",
    "    x = []\n",
    "    y = []\n",
    "    ylow = math.floor(min(mean_scores))-2\n",
    "    yhigh = math.ceil(max(mean_scores))+2\n",
    "    # Create x labels stepwise by 2's.\n",
    "    for n in range(0,readlen+1,2):\n",
    "        x.append(n)\n",
    "    # Create y labels stepwise by 1's.\n",
    "    for n in range(ylow,yhigh):\n",
    "        y.append(n)\n",
    "    # Create the figure, with size and color specifications\n",
    "    plt.figure(num=None, figsize=(16, 8), dpi=80, facecolor='w', edgecolor='k')\n",
    "    # Plot the mean_scores, with a dark blue line.\n",
    "    plt.plot(mean_scores, \"darkblue\")\n",
    "    # Create the x,y tick marks\n",
    "    plt.xticks(x)\n",
    "    plt.yticks(y)\n",
    "    # Label the x,y axes.\n",
    "    plt.xlabel(\"Position of Base in Read\")\n",
    "    plt.ylabel(\"Mean Quality Score\")\n",
    "    # Set the range for the x,y axes.\n",
    "    plt.axis([-1,(readlen+1), ylow, yhigh])\n",
    "    # Title the plot\n",
    "    plt.title(\"Average Quality Scores @ Base Position X\")\n",
    "    # Show the graph grid, then show the figure.\n",
    "    plt.grid(True)\n",
    "    plt.savefig(graph,format='png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index 1 Mean Quality Scores By Nucleotide Position\n",
      "# Base Pos\tMean Quality Score\n",
      "===========\t==================\n",
      "0\t\t30.98\n",
      "1\t\t31.37\n",
      "2\t\t35.29\n",
      "3\t\t35.88\n",
      "4\t\t35.98\n",
      "5\t\t39.35\n",
      "6\t\t39.10\n",
      "7\t\t38.97\n",
      "Read 1 Mean Quality Scores By Nucleotide Position\n",
      "# Base Pos\tMean Quality Score\n",
      "===========\t==================\n",
      "0\t\t31.26\n",
      "1\t\t31.44\n",
      "2\t\t35.33\n",
      "3\t\t35.93\n",
      "4\t\t36.10\n",
      "5\t\t39.57\n",
      "6\t\t39.74\n",
      "7\t\t39.82\n",
      "8\t\t40.01\n",
      "9\t\t40.07\n",
      "10\t\t40.10\n",
      "11\t\t40.12\n",
      "12\t\t40.06\n",
      "13\t\t40.07\n",
      "14\t\t40.07\n",
      "15\t\t40.07\n",
      "16\t\t39.98\n",
      "17\t\t40.01\n",
      "18\t\t40.02\n",
      "19\t\t40.01\n",
      "20\t\t39.98\n",
      "21\t\t40.04\n",
      "22\t\t39.97\n",
      "23\t\t40.02\n",
      "24\t\t39.97\n",
      "25\t\t39.96\n",
      "26\t\t39.90\n",
      "27\t\t39.92\n",
      "28\t\t39.86\n",
      "29\t\t39.91\n",
      "30\t\t39.81\n",
      "31\t\t39.87\n",
      "32\t\t39.85\n",
      "33\t\t39.90\n",
      "34\t\t39.91\n",
      "35\t\t39.90\n",
      "36\t\t39.89\n",
      "37\t\t39.87\n",
      "38\t\t39.88\n",
      "39\t\t39.87\n",
      "40\t\t39.85\n",
      "41\t\t39.84\n",
      "42\t\t39.83\n",
      "43\t\t39.83\n",
      "44\t\t39.74\n",
      "45\t\t39.78\n",
      "46\t\t39.79\n",
      "47\t\t39.79\n",
      "48\t\t39.78\n",
      "49\t\t39.78\n",
      "50\t\t39.78\n",
      "51\t\t39.77\n",
      "52\t\t39.69\n",
      "53\t\t39.72\n",
      "54\t\t39.73\n",
      "55\t\t39.73\n",
      "56\t\t39.67\n",
      "57\t\t39.69\n",
      "58\t\t39.68\n",
      "59\t\t39.67\n",
      "60\t\t39.65\n",
      "61\t\t39.64\n",
      "62\t\t39.59\n",
      "63\t\t39.61\n",
      "64\t\t39.59\n",
      "65\t\t39.58\n",
      "66\t\t39.57\n",
      "67\t\t39.56\n",
      "68\t\t39.54\n",
      "69\t\t39.45\n",
      "70\t\t39.47\n",
      "71\t\t39.46\n",
      "72\t\t39.45\n",
      "73\t\t39.43\n",
      "74\t\t39.41\n",
      "75\t\t39.38\n",
      "76\t\t38.65\n",
      "77\t\t39.06\n",
      "78\t\t39.18\n",
      "79\t\t39.23\n",
      "80\t\t39.32\n",
      "81\t\t39.33\n",
      "82\t\t39.30\n",
      "83\t\t39.29\n",
      "84\t\t39.27\n",
      "85\t\t39.26\n",
      "86\t\t39.22\n",
      "87\t\t39.07\n",
      "88\t\t39.11\n",
      "89\t\t39.12\n",
      "90\t\t39.10\n",
      "91\t\t39.05\n",
      "92\t\t39.02\n",
      "93\t\t38.96\n",
      "94\t\t38.94\n",
      "95\t\t38.94\n",
      "96\t\t38.92\n",
      "97\t\t38.70\n",
      "98\t\t38.75\n",
      "99\t\t38.83\n",
      "100\t\t37.67\n",
      "Index 2 Mean Quality Scores By Nucleotide Position\n",
      "# Base Pos\tMean Quality Score\n",
      "===========\t==================\n",
      "0\t\t30.95\n",
      "1\t\t30.50\n",
      "2\t\t34.04\n",
      "3\t\t35.17\n",
      "4\t\t35.31\n",
      "5\t\t38.49\n",
      "6\t\t37.45\n",
      "7\t\t37.90\n",
      "Read 2 Mean Quality Scores By Nucleotide Position\n",
      "# Base Pos\tMean Quality Score\n",
      "===========\t==================\n",
      "0\t\t30.16\n",
      "1\t\t30.78\n",
      "2\t\t33.78\n",
      "3\t\t34.59\n",
      "4\t\t35.10\n",
      "5\t\t37.95\n",
      "6\t\t38.01\n",
      "7\t\t38.23\n",
      "8\t\t38.44\n",
      "9\t\t38.32\n",
      "10\t\t38.42\n",
      "11\t\t38.45\n",
      "12\t\t38.31\n",
      "13\t\t38.40\n",
      "14\t\t38.25\n",
      "15\t\t38.01\n",
      "16\t\t38.12\n",
      "17\t\t38.26\n",
      "18\t\t38.07\n",
      "19\t\t38.17\n",
      "20\t\t38.02\n",
      "21\t\t38.03\n",
      "22\t\t38.20\n",
      "23\t\t38.22\n",
      "24\t\t38.23\n",
      "25\t\t38.21\n",
      "26\t\t38.18\n",
      "27\t\t38.23\n",
      "28\t\t38.20\n",
      "29\t\t38.15\n",
      "30\t\t38.03\n",
      "31\t\t37.66\n",
      "32\t\t37.57\n",
      "33\t\t37.36\n",
      "34\t\t37.23\n",
      "35\t\t37.14\n",
      "36\t\t37.10\n",
      "37\t\t37.11\n",
      "38\t\t37.08\n",
      "39\t\t37.46\n",
      "40\t\t37.68\n",
      "41\t\t37.85\n",
      "42\t\t37.84\n",
      "43\t\t38.00\n",
      "44\t\t38.02\n",
      "45\t\t38.00\n",
      "46\t\t38.04\n",
      "47\t\t38.00\n",
      "48\t\t37.99\n",
      "49\t\t37.94\n",
      "50\t\t38.02\n",
      "51\t\t38.05\n",
      "52\t\t38.04\n",
      "53\t\t37.84\n",
      "54\t\t37.87\n",
      "55\t\t37.95\n",
      "56\t\t37.96\n",
      "57\t\t37.88\n",
      "58\t\t37.85\n",
      "59\t\t37.78\n",
      "60\t\t37.76\n",
      "61\t\t37.80\n",
      "62\t\t37.81\n",
      "63\t\t37.74\n",
      "64\t\t37.77\n",
      "65\t\t37.70\n",
      "66\t\t37.76\n",
      "67\t\t37.66\n",
      "68\t\t37.61\n",
      "69\t\t37.66\n",
      "70\t\t37.65\n",
      "71\t\t37.59\n",
      "72\t\t37.53\n",
      "73\t\t37.53\n",
      "74\t\t37.53\n",
      "75\t\t37.41\n",
      "76\t\t36.48\n",
      "77\t\t36.88\n",
      "78\t\t37.31\n",
      "79\t\t37.42\n",
      "80\t\t37.54\n",
      "81\t\t37.55\n",
      "82\t\t37.57\n",
      "83\t\t37.48\n",
      "84\t\t37.39\n",
      "85\t\t37.40\n",
      "86\t\t37.33\n",
      "87\t\t37.34\n",
      "88\t\t37.30\n",
      "89\t\t37.23\n",
      "90\t\t37.19\n",
      "91\t\t37.10\n",
      "92\t\t37.07\n",
      "93\t\t37.03\n",
      "94\t\t36.85\n",
      "95\t\t36.87\n",
      "96\t\t36.87\n",
      "97\t\t36.80\n",
      "98\t\t36.54\n",
      "99\t\t36.62\n",
      "100\t\t34.66\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "echo \"Index 1 Mean Quality Scores By Nucleotide Position\"\n",
    "cat 1294*R2*.txt\n",
    "echo \"Read 1 Mean Quality Scores By Nucleotide Position\"\n",
    "cat 1294*R1*.txt\n",
    "echo \"Index 2 Mean Quality Scores By Nucleotide Position\"\n",
    "cat 1294*R3*.txt  \n",
    "echo \"Read 2 Mean Quality Scores By Nucleotide Position\"\n",
    "cat 1294*R4*.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Read 1**  \n",
    "![Read1](1294_S1_L008_R1_001.png)  \n",
    "**Index 1**  \n",
    "![Index1](1294_S1_L008_R2_001.png)  \n",
    "**Read 2**  \n",
    "![Read2](1294_S1_L008_R4_001.png)  \n",
    "**Index 2**  \n",
    "![Index2](1294_S1_L008_R3_001.png)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2b) Phred Quality-Score Cut-off:  \n",
    "---  \n",
    "> A good quality-score cutoff often depends on the particular applicatation. A quality-score of 20 or higher is often considered usable data, because this means that only 1% of bases are likely to be erroneous. In other cases, we may desire greater certainty, in which case a quality-score cutoff of 30, meaning only 0.1% of bases are likely to be erroneous, can be a useful target. As you can see from the graph below, quality scores below 20 increase in error rates precipitously. For this reason, sequences with a quality-score below 20 are considered unreliable.   \n",
    "\n",
    "![QscoreErrorRate](PhredQscoreErrors.png)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2c) Number of Indexes Containing an \"N\" undetermined base call:  \n",
    "---  \n",
    "**The following shell commands were used to retrieve this information:**  \n",
    ">zcat /projects/bgmp/shared/2017_sequencing/1294_S1_L008_R2_001.fastq.gz | sed -n '2~4p' | grep -cE 'N' > index1Ncount.txt  \n",
    ">zcat /projects/bgmp/shared/2017_sequencing/1294_S1_L008_R3_001.fastq.gz | sed -n '2~4p' | grep -cE 'N' > index2Ncount.txt  \n",
    "\n",
    "_Index 1:  \n",
    "**3,976,613** indexes contained an \"N\"_  \n",
    "_Index 2:  \n",
    "**3,328,051** indexes contained an \"N\"_\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 2  \n",
    "---  \n",
    "> Generate a strategy for developing an algorithm designed to demultiplex files and report index-hopping. That is, given four files (2 w/ biological reads, 2 w/ index reads) and known indexes, sort reads by index, outputting one forward file and one reverse file per index, plus a pair of files for unknown indexes. Additionally, your algorithm should report the number of properly matched indexes (per index) and the level of index hopping observed.  \n",
    "\n",
    "1. Define the Problem:  \n",
    "    The problem is index hopping. By chance, a certain level of index hopping is going to occur. These reads amount to\n",
    "    lost data if their true identities cannot be resolved. Or, if reads are included with ambiguous identities, the \n",
    "    alignment process becomes complicated, if not confounded. The simplest approach would be to separate reads whose \n",
    "    indexes contain any ambiguity from those that have clearly matched indexes. We also need to keep tabs on all four\n",
    "    files simultaneously to both determine properly matched indexes and to print each read to its appropriate file. To\n",
    "    accurately count index hopping, it would be necessary to identify which index mismatches were a result of\n",
    "    sequencing error and which were a result of actual index hopping. This complicates the problem, as we'll need\n",
    "    to track each character in each index for rates of similarity to classify each mismatch into 3 categories: as the\n",
    "    result of sequencing error, index hopping, or remaining ambiguous. Statistical methods would need to be devised to \n",
    "    validate any such process of categorization.\n",
    "\n",
    "2. Determine/Describe What Output Would Be Informative:  \n",
    "    The forward and reverse output files of properly matched indices would contain the genomic information of interest.   \n",
    "    It would also be useful to know how many mismatches are due to sequencing error, how many are due to index  \n",
    "    hopping, and how many are unresolvable mismatches.   \n",
    "3. Write Examples (Unit Tests!):  \n",
    "    i. Include four properly formatted input fastq files.  \n",
    ">All Files Associated With a Unit Test End With '.fq' In the Github Folder   \n",
    "    \n",
    "    ii. Include the appropriate number of properly formatted output fastq files.  \n",
    "    \n",
    "            I created 4 indexes, so there are 4 forward read and 4 reverse read files, \n",
    "            plus 2 files for mismatches.  \n",
    "4. Develop Your Algorithm Using Pseudocode:  \n",
    "    I. Initialize a dictionary containing the 24 valid indices as keys, each assigned its own unique 2 string character\n",
    "    IDs as values, to be used to name and identify output files.  \n",
    "    \n",
    "    II. Initialize an 'index hopping' counter and a 'positive match' counter.\n",
    "    \n",
    "    II. Open the 4 files for reading.\n",
    "    \n",
    "    III. For each read, check if Index 1 is in the dictionary and equal to Index 2.\n",
    ">If yes, look up the index's 2 character ID in the dictionary, open 2 files for appending, with that ID +\n",
    "['for' OR 'rev'] to indicate forward and reverse strands. Write the current reads from the forward and reverse\n",
    "files to their own newly created file.  \n",
    "  \n",
    ">If no, call a comparison function to determine index similarity between index 1 and index 2. If they are\n",
    "sufficiently similar, open the 2 mismatch files for appending, and append reads 1 and 2. If they are\n",
    "sufficiently dissimilar, open the 2 mismatch files for appending, and append reads 1 and 2, then increment the \n",
    "index hopping counter.    \n",
    "    \n",
    "5. Determine High-Level Functions:  \n",
    "Description  \n",
    ">A high-level function to determine index similarity by comparing base-by-base between the two indices. If the 2\n",
    " indices are dissimilar, then return a Boolean True that will trigger the index-hopping counter, in-script, to\n",
    " increment.  \n",
    " \n",
    "Function Headers \n",
    ">INPUT: 2 strings of nucleotides. OUTPUT: A Boolean  \n",
    "\n",
    "Test Examples for Individual Functions  \n",
    ">'ACTGACTG' compared to 'ACTGACTA' returns FALSE  \n",
    "'ACTGACTG' compared to 'ACTGACCA' returns FALSE  \n",
    "'ACTGACTG' compared to 'ACCAGTCA' returns TRUE  \n",
    "'ACTGACTG' compared to 'CAGACAGA' returns TRUE  \n",
    "2 or less differences returns FALSE.  \n",
    "2 or less similarities returns TRUE.  \n",
    "\n",
    "Return Statement  \n",
    ">The function returns TRUE or FALSE, which in the script determines whether to increment the index-hopping counter.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
